hpedockerplugin/backend_orchestrator.py

```diff
@@ -88,19 +88,36 @@ def initialize_manager_objects(self, host_config, backend_configs):
 
     def get_volume_backend_details(self, volname):
         LOG.info('Getting details for volume : %s ' % (volname))
-        current_backend = DEFAULT_BACKEND_NAME
 
         if volname in self.volume_backends_map:
             current_backend = self.volume_backends_map[volname]
             LOG.debug(' Returning the backend details from cache %s , %s'
                       % (volname, current_backend))
             return current_backend
-
-        vol = self.etcd_util.get_vol_byname(volname)
-        if vol is not None and 'backend' in vol:
-            current_backend = vol['backend']
-
-        return current_backend
+        else:
+            return self.add_cache_entry(volname)
+
+    def add_cache_entry(self, volname):
+        # Using this style of locking
+        # https://docs.python.org/3/library/threading.htmls#using-locks-conditions-and-semaphores-in-the-with-statement
+        self.volume_backend_lock.acquire()
+        try:
+            vol = self.etcd_util.get_vol_byname(volname)
+            if vol is not None and 'backend' in vol:
+                current_backend = vol['backend']
+                # populate the volume backend map for caching
+                LOG.debug(' Populating cache %s, %s '
+                          % (volname, current_backend))
+                self.volume_backends_map[volname] = current_backend
+                return current_backend
+            else:
+                # throw an exception for the condition
+                # where the backend can't be read from volume
+                # metadata in etcd
+                LOG.info(' vol obj read from etcd : %s' % vol)
```

Comment:
Should this be:
"vol obj is either not found in ETCD or it doesn't have backend set"